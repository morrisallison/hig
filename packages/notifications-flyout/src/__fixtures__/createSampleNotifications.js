import React from "react";
import TextLink from "@hig/text-link";
import Timestamp from "@hig/timestamp";

export default function createSampleNotifications({ baseDate = new Date() }) {
  const updatedDate1 = baseDate.setMinutes(baseDate.getMinutes() - 3);
  const updatedDate2 = baseDate.setHours(baseDate.getHours() - 2);
  const updatedDate3 = baseDate.setHours(baseDate.getHours() - 24);
  const updatedDate4 = baseDate.setMinutes(baseDate.getMinutes() - 20);

  return [
    {
      id: "featured",
      featured: true,
      unread: true,
      timestamp: <Timestamp timestamp={new Date(updatedDate1).toISOString()} />,
      content: (
        <div>
          <p>
            <b>Your subscription expires May 5</b>
          </p>
          <p>
            Maya
            <br />
            Media & Entertainment Collection
            <br />
            Product Design Collection
            <br />2 more
          </p>
          <p>
            <TextLink
              link="https://github.com/Autodesk/hig"
              onClick={() => {
                console.log("notifications id 1");
              }}
            >
              Manage renewal
            </TextLink>
          </p>
        </div>
      )
    },
    {
      id: "0",
      unread: true,
      timestamp: <Timestamp timestamp={new Date(updatedDate1).toISOString()} />,
      type: "primary",
      content: (
        <div>
          <p>
            <b>Your subscription expires May 5</b>
          </p>
          <p>
            Maya
            <br />
            Media & Entertainment Collection
            <br />
            Product Design Collection
            <br />2 more
          </p>
          <p>
            <TextLink
              link="https://github.com/Autodesk/hig"
              onClick={() => {
                console.log("notifications id 1");
              }}
            >
              Manage renewal
            </TextLink>
          </p>
        </div>
      )
    },
    {
      id: "1",
      unread: true,
      timestamp: <Timestamp timestamp={new Date(updatedDate2).toISOString()} />,
      type: "success",
      content: (
        <div>
          <p>
            <b>Your subscription expires April 20</b>
          </p>
          <p>
            AutoCAD
            <br />
            Architecture Construction Engineering Collection
            <br />
            Product Design Collection
            <br />
          </p>
          <p>
            <TextLink
              link="https://github.com/Autodesk/hig"
              onClick={() => {
                console.log("notifications id 2");
              }}
            >
              Manage renewal
            </TextLink>
          </p>
        </div>
      )
    },
    {
      id: "3",
      unread: true,
      timestamp: <Timestamp timestamp={new Date(updatedDate3).toISOString()} />,
      type: "error",
      content: (
        <div>
          <p>
            <b>Your subscription expires June 15</b>
          </p>
          <p>
            AutoCAD
            <br />
            Architecture Construction Engineering Collection
            <br />
            HIG
            <br />
          </p>
          <p>
            <TextLink
              link="https://github.com/Autodesk/hig"
              onClick={() => {
                console.log("notifications id 2");
              }}
            >
              Manage Renewal
            </TextLink>
          </p>
        </div>
      )
    },
    {
      id: "4",
      unread: true,
      timestamp: <Timestamp timestamp={new Date(updatedDate4).toISOString()} />,
      type: "warning",
      content: (
        <div>
          <p>
            <b>Your subscription expires June 15</b>
          </p>
          <p>A new version of Autodesk Revit is available for download.</p>
          <p>
            <TextLink
              link="https://github.com/Autodesk/hig"
              onClick={() => {
                console.log("notifications id 2");
              }}
            >
              Manage renewal
            </TextLink>
          </p>
        </div>
      )
    }
  ];
}
