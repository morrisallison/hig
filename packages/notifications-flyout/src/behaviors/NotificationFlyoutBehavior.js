/* eslint-disable react/no-unused-prop-types */

import { Component } from "react";
import PropTypes from "prop-types";
import { polyfill } from "react-lifecycles-compat";

import parseNotifications from "./parseNotifications";

/** @typedef {import("./parseNotifications").Input} RawNotifications */
/** @typedef {import("./parseNotifications").ParsedNotification} ParsedNotification */
/** @typedef {import("./NotificationBehavior").Payload} NotificationBehaviorPayload */

/**
 * @typedef {Object} Payload
 * @property {Function} dismissNotification
 * @property {ParsedNotification[]} notifications
 * @property {boolean} showUnreadCount
 * @property {number} unreadCount
 */

/**
 * @typedef {Object} Props
 * @property {function(Payload): import("react").ReactElement} [children]
 * @property {import("./parseNotifications").Input} [notifications]
 * @property {number} [unreadCount]
 */

/**
 * @typedef {Object} State
 * @property {string[]} dismissedNotifications An array of IDs of notifications that have been dismissed
 * @property {ParsedNotification[]} notifications
 * @property {string[]} readNotifications An array of IDs of notifications that have been read
 */

class NotificationFlyoutBehavior extends Component {
  static propTypes = {
    /** Render prop */
    children: PropTypes.func,
    /** Rendered notifications */
    notifications: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.node, PropTypes.shape({})])
      )
    ]),
    /** When provided, it overrides the derived unread notification count */
    unreadCount: PropTypes.number
  };

  static defaultProps = {
    title: "Notifications"
  };

  /** @type {State} */
  state = {
    dismissedNotifications: [],
    notifications: [],
    readNotifications: []
  };

  /**
   * @param {Props} nextProps
   * @returns {State | null}
   */
  static getDerivedStateFromProps(nextProps) {
    return {
      notifications: parseNotifications(nextProps.notifications)
    };
  }

  /**
   * @returns {ParsedNotification[]}
   */
  getNotifications() {
    const {
      dismissedNotifications,
      notifications,
      readNotifications
    } = this.state;

    const updateReadStatus = ({ id, unread, ...otherProps }) => ({
      id,
      unread: unread && !readNotifications.includes(id),
      ...otherProps
    });
    const isNotDismissed = ({ id }) => !dismissedNotifications.includes(id);

    return notifications.map(updateReadStatus).filter(isNotDismissed);
  }

  /** @returns {number} */
  getUnreadCount() {
    const { unreadCount: controlledUnreadCount } = this.props;

    return controlledUnreadCount !== undefined
      ? controlledUnreadCount
      : this.deriveUnreadCount();
  }

  /**
   * @param {ParsedNotification[]} notifications
   */
  deriveUnreadCount() {
    return this.getNotifications().reduce(
      (count, { unread }) => (unread ? count + 1 : count),
      0
    );
  }

  /** @type {Props} */
  props;

  markAllNotificationsRead() {
    const notifications = this.getNotifications();
    const { readNotifications } = this.state;
    const nextRead = notifications.reduce((result, notification) => {
      const { id } = notification;

      if (!result.includes(id)) result.push(id);

      return result;
    }, readNotifications.slice());

    this.setState({ readNotifications: nextRead });
  }

  /**
   * Action to dismiss a notification
   * @param {string} id
   */
  dismissNotification = id => {
    this.setState({
      dismissedNotifications: this.state.dismissedNotifications.concat(id)
    });
  };

  /**
   * Handler for when the flyout opens
   */
  handleClose = () => {
    this.markAllNotificationsRead();
  };

  /**
   * @returns {import("react").ReactElement}
   */
  render() {
    const { dismissNotification, handleClose } = this;
    const notifications = this.getNotifications();
    const unreadCount = this.getUnreadCount();
    const showUnreadCount = unreadCount > 0;

    return this.props.children({
      dismissNotification,
      handleClose,
      notifications,
      showUnreadCount,
      unreadCount
    });
  }
}

export default polyfill(NotificationFlyoutBehavior);
