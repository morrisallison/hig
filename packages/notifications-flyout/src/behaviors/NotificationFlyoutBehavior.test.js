import React from "react";
import { mount } from "enzyme";

import createSampleNotifications from "../__fixtures__/createSampleNotifications";
import NotificationFlyoutBehavior from "./NotificationFlyoutBehavior";

describe("notification-flyout/behaviors/NotificationFlyoutBehavior", () => {
  const baseDate = new Date("2018-08-20T18:58:55.441Z");
  const children = jest.fn();

  function mountBehavior() {
    return mount(
      <NotificationFlyoutBehavior
        notifications={createSampleNotifications({ baseDate })}
      >
        {children}
      </NotificationFlyoutBehavior>
    );
  }

  beforeEach(() => {
    children.mockReturnValue(<div />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders", () => {
    mountBehavior();
  });

  describe("render prop payload", () => {
    let payload;
    let wrapper;

    beforeEach(() => {
      wrapper = mountBehavior();
      payload = children.mock.calls[0][0];
    });

    it("provides a payload to the `children` render prop", () => {
      expect(payload).toMatchObject({
        dismissNotification: expect.any(Function),
        handleClose: expect.any(Function),
        notifications: expect.any(Array),
        showUnreadCount: true,
        unreadCount: 5
      });
      expect(payload.notifications).toMatchSnapshot();
    });

    it("normalizes notifications", () => {
      expect(payload.notifications).toMatchSnapshot();
    });

    describe("dismissNotification", () => {
      function expectToHaveNotificationTimes(length) {
        const { notifications } = children.mock.calls.slice(-1)[0][0];

        expect(notifications).toHaveLength(length);
      }

      it("filters out the notification of the given ID", () => {
        const { dismissNotification } = payload;

        expectToHaveNotificationTimes(5);
        dismissNotification("featured");
        expectToHaveNotificationTimes(4);
        dismissNotification("featured");
        expectToHaveNotificationTimes(4);
      });
    });

    describe("handleClose", () => {
      function expectToHaveUnreadNotificationTimes(length) {
        const { notifications } = children.mock.calls.slice(-1)[0][0];
        const unreadNotifications = notifications.filter(
          ({ unread }) => unread
        );

        expect(unreadNotifications).toHaveLength(length);
      }

      it("marks all notifications as read", () => {
        const { handleClose } = payload;

        expectToHaveUnreadNotificationTimes(5);
        handleClose();
        expectToHaveUnreadNotificationTimes(0);
      });

      it("updates the unread count", () => {
        const { handleClose } = payload;

        expect(children).toHaveBeenLastCalledWith(
          expect.objectContaining({ unreadCount: 5 })
        );

        handleClose();

        expect(children).toHaveBeenLastCalledWith(
          expect.objectContaining({ unreadCount: 0 })
        );
      });
    });

    describe("unreadCount", () => {
      it("is derived from the given notifications", () => {
        expect(payload.unreadCount).toEqual(5);
      });

      describe("when controlled", () => {
        it("it references the `unreadCount` prop", () => {
          const unreadCount = 3;

          wrapper.setProps({ unreadCount });

          expect(children).toHaveBeenLastCalledWith(
            expect.objectContaining({ unreadCount })
          );
        });
      });
    });
  });
});
