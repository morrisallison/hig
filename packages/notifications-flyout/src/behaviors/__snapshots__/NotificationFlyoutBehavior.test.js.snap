// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notification-flyout/behaviors/NotificationFlyoutBehavior render prop payload normalizes notifications 1`] = `
Array [
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires May 5
        </b>
      </p>
      <p>
        Maya
        <br />
        Media & Entertainment Collection
        <br />
        Product Design Collection
        <br />
        2 more
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "featured": true,
    "id": "featured",
    "key": "notification-0",
    "timestamp": <Timestamp
      timestamp="2018-08-18T14:09:55.441Z"
    />,
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires May 5
        </b>
      </p>
      <p>
        Maya
        <br />
        Media & Entertainment Collection
        <br />
        Product Design Collection
        <br />
        2 more
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "0",
    "key": "notification-1",
    "timestamp": <Timestamp
      timestamp="2018-08-18T14:09:55.441Z"
    />,
    "type": "primary",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires April 20
        </b>
      </p>
      <p>
        AutoCAD
        <br />
        Architecture Construction Engineering Collection
        <br />
        Product Design Collection
        <br />
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "1",
    "key": "notification-2",
    "timestamp": <Timestamp
      timestamp="2018-08-18T12:09:55.441Z"
    />,
    "type": "success",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires June 15
        </b>
      </p>
      <p>
        AutoCAD
        <br />
        Architecture Construction Engineering Collection
        <br />
        HIG
        <br />
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage Renewal
        </TextLink>
      </p>
    </div>,
    "id": "3",
    "key": "notification-3",
    "timestamp": <Timestamp
      timestamp="2018-08-17T12:09:55.441Z"
    />,
    "type": "error",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires June 15
        </b>
      </p>
      <p>
        A new version of Autodesk Revit is available for download.
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "4",
    "key": "notification-4",
    "timestamp": <Timestamp
      timestamp="2018-08-17T11:49:55.441Z"
    />,
    "type": "warning",
    "unread": true,
  },
]
`;

exports[`notification-flyout/behaviors/NotificationFlyoutBehavior render prop payload provides a payload to the \`children\` render prop 1`] = `
Array [
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires May 5
        </b>
      </p>
      <p>
        Maya
        <br />
        Media & Entertainment Collection
        <br />
        Product Design Collection
        <br />
        2 more
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "featured": true,
    "id": "featured",
    "key": "notification-0",
    "timestamp": <Timestamp
      timestamp="2018-08-19T16:32:55.441Z"
    />,
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires May 5
        </b>
      </p>
      <p>
        Maya
        <br />
        Media & Entertainment Collection
        <br />
        Product Design Collection
        <br />
        2 more
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "0",
    "key": "notification-1",
    "timestamp": <Timestamp
      timestamp="2018-08-19T16:32:55.441Z"
    />,
    "type": "primary",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires April 20
        </b>
      </p>
      <p>
        AutoCAD
        <br />
        Architecture Construction Engineering Collection
        <br />
        Product Design Collection
        <br />
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "1",
    "key": "notification-2",
    "timestamp": <Timestamp
      timestamp="2018-08-19T14:32:55.441Z"
    />,
    "type": "success",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires June 15
        </b>
      </p>
      <p>
        AutoCAD
        <br />
        Architecture Construction Engineering Collection
        <br />
        HIG
        <br />
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage Renewal
        </TextLink>
      </p>
    </div>,
    "id": "3",
    "key": "notification-3",
    "timestamp": <Timestamp
      timestamp="2018-08-18T14:32:55.441Z"
    />,
    "type": "error",
    "unread": true,
  },
  Object {
    "content": <div>
      <p>
        <b>
          Your subscription expires June 15
        </b>
      </p>
      <p>
        A new version of Autodesk Revit is available for download.
      </p>
      <p>
        <TextLink
          link="https://github.com/Autodesk/hig"
          onClick={[Function]}
          type="primary"
        >
          Manage renewal
        </TextLink>
      </p>
    </div>,
    "id": "4",
    "key": "notification-4",
    "timestamp": <Timestamp
      timestamp="2018-08-18T14:12:55.441Z"
    />,
    "type": "warning",
    "unread": true,
  },
]
`;
